### ISSUE TITLE
Regression: Deprecation of .onConflict in InsertQueryBuilder

### ISSUE THREAD

**Issue by konqi** [typeorm/typeorm#8124]
## Issue Description
Since this commit https://github.com/typeorm/typeorm/commit/a1c6bb0092b63fa808d9f836936238a7977a8027 by @imnotjames the `.onConflict`-method of the InsertQueryBuilder is marked as deprecated and refers to `.onUpdate` as an alternative.

However `.onUpdate` does not provide the possibility to define an additional WHERE condition. In my current project we have an upsert function that will attempt an insert and fallback to an update. The data to be upserted is possibly outdated so we pass a WHERE condition in the `.onConflict` that checks a timestamp.

Example:
```
.onConflict(
            `("${conflictingColumn}") DO UPDATE
                SET (${columnName}) = (${value})
                WHERE timestamp < ${eventTimestamp}`
        )
```

Possible remedies for the situation are:
- try insert, catch conflict, try update; but that would require a wrapping transaction and locks on the table, that I don't necessarily want
- let typeorm generate the insert and append the conflict handle by hand (which feels super hacky)

So for now I'm stuck with the deprecated `.onConflict` and I fear it will be removed without an alternative for this use case. So my questions is: Is this use case so special or did I miss something? Am I doing something completely wrong? If not, I would ask to no deprecate the `.onConflict` method without an alternative.

### Relevant Database Driver(s)

| DB Type              | Reproducible  |
| ---                  | ---           |
| `aurora-data-api`    | no            |
| `aurora-data-api-pg` | no            |
| `better-sqlite3`     | no            |
| `cockroachdb`        | no            |
| `cordova`            | no            |
| `expo`               | no            |
| `mongodb`            | no            |
| `mysql`              | no            |
| `nativescript`       | no            |
| `oracle`             | no            |
| `postgres`           | yes            |
| `react-native`       | no            |
| `sap`                | no            |
| `sqlite`             | no            |
| `sqlite-abstract`    | no            |
| `sqljs`              | no            |
| `sqlserver`          | no            |

### Are you willing to resolve this issue by submitting a Pull Request?

 - ✖️ Yes, I have the time, and I know how to start.
 - ✅ Yes, I have the time, but I don't know how to start. I would need guidance.
 - ✖️ No, I don’t have the time, but I can support (using donations) development.
 - ✖️ No, I don’t have the time and I’m okay to wait for the community / maintainers to resolve this issue.

**Comment by wilson208** [comment]
@konqi I had a need for similar functionality in MySQL and SQLite and opened a PR with a potential solution, it doesn't go as far as `orUpdate` accepting an UpdateQueryBuilder, but lets an expression be passed like so:

MySQL
```
.orUpdate([{column: 'views', expression: 'views + VALUES(views)'}], ['feedItemId', 'date'])
```

Postgres / SQLite like so:
```
 .orUpdate([{column: 'views', expression: 'views + EXCLUDED.views'}], ['feedItemId', 'date'])
```

https://github.com/typeorm/typeorm/pull/8127

If you think this would be a potential solution for you, or see any issues I would appreciate the feedback, would love to get my PR merged / released because I am currently targetting my fork in my project because there is no workaround for mysql, `onConflict`, even though it is now deprecated, only works with Postgres.

**Comment by konqi** [comment]
@wilson208 I think it would be viable workaround, not a full solution yet. In an ideal world we don't write (potentially dangerous/unsafe) partial SQL statements. Your change makes it possible to use `.orUpdate`, yes - but to honest, I would stick with `.onConflict` until a full solution comes around. I do understand this is not an option for you given your use of mysql.

**Comment by smcgivern** [comment]
`.onConflict` also hasn't been replaced in another sense - I don't see a way to do `ON CONFLICT (a, b, ...) DO NOTHING` any more.

If we provide `.orIgnore`, we just get `ON CONFLICT DO NOTHING` with no way to set the columns: https://github.com/typeorm/typeorm/blob/daf1b47a609cb5090b71e537668ee115a98b1dc9/src/query-builder/InsertQueryBuilder.ts#L487

And if we do `.orUpdate`, we always seem to send up with `ON CONFLICT ... DO UPDATE ...`, with no ability to say `DO NOTHING`.

I know that `.onConflict` is just deprecated, not removed, but is there any issue tracking that?

**Comment by barraponto** [comment]
I agree with the points here. Also, there's some freedom to `.onConflict` allowing some arbitrary query.

Mainly, though, I wanted to see support for a WHERE condition, just like there is an `index_predicate` option.
We could have it in [InsertOrUpdateOptions](https://github.com/typeorm/typeorm/blob/master/src/query-builder/InsertOrUpdateOptions.ts) and have the query builder apply it if present.

**Comment by ian-axelrod** [comment]
Please do support `WHERE` conditions for `orUpdate`. Cheers.

**Comment by zbarbuto** [comment]
I believe this is still relevant?

**Comment by Juraj-Masiar** [comment]
So what's the recommended approach here?
I've just upgraded to version 1.0 which doesn't have `onConflict` anymore, but `onUpdate` doesn't allow custom query text, right?
I need to increment value by one:
```ts
.onConflict(`("domain") DO UPDATE SET "requested" = "missing_img"."requested" + 1`)
```
So I need to remove the `createQueryBuilder` and rewrite it in native SQL, correct?

**Comment by mprisehk** [comment]
Another thing that .onConflict could be used for, which is also no longer possible now: using calculated values when a conflict happens. For my use case, I'd like to do the following:
```sql
ON CONFLICT ("inventoryId")
DO UPDATE SET "quantity" = "quantity" + EXCLUDED."quantity"
```
However the .orUpdate() syntax only allows this:
```ts
.insert()
.values({
  inventoryId,
  quantity,
})
.orUpdate(
  ['quantity'],
  ['inventoryId'],
)
```
which results in:
```sql
ON CONFLICT ("inventoryId")
DO UPDATE SET "quantity" = EXCLUDED."quantity"
```
This has been reported earlier in issue #8125, which has unfortunately been closed due to staleness.
